README.md

```diff
@@ -54,10 +54,15 @@ Other supported formats are listed below.
   * true - Server certificate is not checked. Default is true if encrypt is not specified. If trust server certificate is true, driver accepts any certificate presented by the server and any host name in that certificate. In this mode, TLS is susceptible to man-in-the-middle attacks. This should be used only for testing.
 * `certificate` - The file that contains the public key certificate of the CA that signed the SQL Server certificate. The specified certificate overrides the go platform specific CA certificates.
 * `hostNameInCertificate` - Specifies the Common Name (CN) in the server certificate. Default value is the server host.
-* `ServerSPN` - The kerberos SPN (Service Principal Name) for the server. Default is MSSQLSvc/host:port.
+* `ServerSPN` - The Kerberos SPN (Service Principal Name) for the server. Default is MSSQLSvc/host:port.
 * `Workstation ID` - The workstation name (default is the host name)
 * `ApplicationIntent` - Can be given the value `ReadOnly` to initiate a read-only connection to an Availability Group listener. The `database` must be specified when connecting with `Application Intent` set to `ReadOnly`. 
-
+* `FedAuth` - The federated authentication scheme to use. See below for additional setup requirements.
+  * `ActiveDirectoryApplication` - authenticates using an Azure Active Directory application client ID and client secret or certificate. Set the `user` to `client-ID@tenant-ID` and the `password` to the client secret. If using client certificates, provide the path to the PKCS#12 file containing the certificate and RSA private key in the `ClientCertPath` parameter, and set the `password` to the value needed to open the PKCS#12 file.
+  * `ActiveDirectoryMSI` - authenticates using the managed service identity (MSI) attached to the VM, or a specific user-assigned identity if a client ID is specified in the `user` field.
```

Comment:
next step wish : 
avoid specifying the clientID.
useMSI + identityName or resourceID would be nicer, as the clientID is only known after deploying the MSI. 
the resourceID/name is known before deployment. this helps in configuring multiple environment statically.

Reply:
I think to look up the clientID you would need Azure resource manager (portal) access, or at least Azure AD, and I think you might need an initial authentication context to be able to do that. So you would have a chicken-and-egg situation - to get the client ID to use as part of your credentials, you'd need credentials. You may have more luck working on a deploy-time solution to look up the client ID though.

When I've been deploying this I was using the Kubernetes AAD pod identity controller, and the client ID is put in there. We deploy using Azure DevOps pipelines, and the deploy pipeline is able to do the lookup using credentials available to the pipeline.

Reply:
no, the IMDS endpoint on the host has an endpoint to get the token from the MSI resource ID. we added support for that in ADAL just a few weeks ago : 
https://github.com/Azure/go-autorest/pull/544

Reply:
"When I've been deploying this I was using the Kubernetes AAD pod identity controller, and the client ID is put in there. We deploy using Azure DevOps pipelines, and the deploy pipeline is able to do the lookup using credentials available to the pipeline."
that means you have dynamic configuration, fetching an ID at deployment time.
I would rather be able to avoid this, and keep the configuration static :). 
not a necessity for this PR, just providing my point of view based on real case constraints we have

Reply:
Oh nice! I had not seen that!

Reply:
LOL but I see now the PR is failing in CI because of other changes in the go-autorest library: we may not be able to use the more recent library and maintain the Go version compatibility (1.8+) that was previously possible.

Reply:
I'll send a small PR to just add the context to the tokenprovider func.
that would make the current ode viable, while waiting for this to land.
Thanks a lot for this work though. again, happy to help land it if needed

Reply:
I put in a workaround in the CI builds to check out the last tag that builds from the ADAL project - I think it actually includes the change you've mentioned above, though.

Reply:
And I created [an issue](https://github.com/Azure/go-autorest/issues/555) in the go-autorest project: I'm not sure if they consider supporting Go prior to 1.13 important, though.

Reply:
the issue mentioned above is fixed in latest version of autorest. it should be good to go.